### PR TITLE
Enhance Viber Adapter with Tracking Data and NLP

### PR DESCRIPTION
**Pull request by DimitriosGkegkas**
## Below you can see a list of the files I changed and why
1. ### examples/viber-demo-bot/src/
- modules/getStarted/actions.ts
This is the viber example to test that everything is working properly.
- bot.ts
Here in the example you can add a Wit.ai token in order to test the NLP functionality. 

2. ### packages/framework/lib/
- adapter.ts
Change the interface of the text handler to add optional data property as well as text
- text.ts
The data property handled by text handler (if exists) should be past as extra payload parameter.

3. ### packages/viber-adapter/lib/
- adapter.ts
Adapter now has an extra nlp property which is a Wit object that can handle wit.ai API requests.
this wit property is passed as parameter in all function needed in order to send the nlp meaning to the textHandler
`Future work to add the functions as adaptor's methods in order to use the nlp property inside the adaptor class`
- nlpWit.ts
This is a new class in order to perform Wit.ai API request.
The only functionality it has for now it is to get the meaning of a text.

### CODE REVIEW

**Comment by DimitriosGkegkas** [comment]
Rebased to main and cleaned commits @chrispanag

**Comment by chrispanag** [comment]
@DimitriosGkegkas will check it thoroughly today's afternoon

**Comment by DimitriosGkegkas** [comment]
@chrispanag Hi there! Do we have any update on that?

File: packages/framework/lib/handlers/text.ts [hunk 1]
```diff
@@ -35,12 +35,15 @@ export default function textHandlerFactory<U extends User<any>>(
 ) {
     function textHandler(
         this: Bot<U>,
-        message: { text: string },
+        message: { text: string; data?: { [key: string]: any } },
```
Comment (chrispanag) [inline packages/framework/lib/handlers/text.ts:38]: why do we need that extra data field? what are we inputting to this? Couldn't you just pass the nlp object to the action?

File: packages/framework/lib/adapter.ts [hunk 2]
```diff
@@ -14,7 +14,11 @@ export interface IRouters {
 
 export interface EbonyHandlers<U> {
     attachment?: (user: U, attachment: GenericAttachment) => Promise<any>;
-    text?: (message: { text: string }, nlp: WitNLP | undefined, user: U) => Promise<any>;
+    text?: (
+        message: { text: string; data?: { [key: string]: any } },
```
Comment (chrispanag) [inline packages/framework/lib/adapter.ts:18]: See question on text.ts file

File: packages/viber-adapter/lib/nlpWit.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            options.method === 'GET' && options.params
+                ? this.url(apiKey) + options.params
+                : this.url(apiKey);
+
+        return fetch(requestURL, options)
+            .then(checkStatus)
+            .then((data: any) => data as WitNLP)
+            .catch((err: any) => undefined);
+    }
+    public async Meaning(message: string): Promise<WitNLP | undefined> {
```
Comment (chrispanag) [inline packages/viber-adapter/lib/nlpWit.ts:36]: class methods should not start with a capital letter. Only classes. 

`Meaning` -> `meaning` (find other places in the code that do this)

**Review by chrispanag — CHANGES_REQUESTED** [review]
Let's start with these...

File: packages/framework/lib/handlers/text.ts (hunk 1 above)
Comment (DimitriosGkegkas) [reply packages/framework/lib/handlers/text.ts:38]: The data field is needed for the tracking data enhancement.
We have the ability to pass a data object: { [key: string]: any } regardless if it is a post back tracking_data or not.
This way each function that response to a text now has the information of the text written by the user (or send by a button press)  as well as the tracking data we provided on previews calls.

File: packages/framework/lib/adapter.ts (hunk 2 above)
Comment (DimitriosGkegkas) [reply packages/framework/lib/adapter.ts:18]: Same as answer on text.ts file

File: packages/viber-adapter/lib/nlpWit.ts (hunk 3 above)
Comment (DimitriosGkegkas) [reply packages/viber-adapter/lib/nlpWit.ts:36]: Of course. I am on it!

**Comment by DimitriosGkegkas** [comment]
This request is closed. 
This features will be re developed and I will open a new request.

**Pull request closed without merging**